coderd/cryptokeys/dbkeycache.go

```diff
@@ -34,18 +33,34 @@ type DBCache struct {
 	closed       bool
 }
 
-type DBCacheOption func(*DBCache)
+type DBCacheOption func(*dbCache)
 
 func WithDBCacheClock(clock quartz.Clock) DBCacheOption {
-	return func(d *DBCache) {
+	return func(d *dbCache) {
 		d.clock = clock
 	}
 }
 
-// NewDBCache creates a new DBCache. Close should be called to
+// NewSigningCache creates a new DBCache. Close should be called to
 // release resources associated with its internal timer.
-func NewDBCache(logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*DBCache)) *DBCache {
-	d := &DBCache{
+func NewSigningCache(logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*dbCache)) (SigningKeycache, error) {
+	if !isSigningKeyFeature(feature) {
+		return nil, ErrInvalidFeature
+	}
+
+	return newDBCache(logger, db, feature, opts...), nil
+}
+
+func NewEncryptionCache(logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*dbCache)) (EncryptionKeycache, error) {
+	if !isEncryptionKeyFeature(feature) {
+		return nil, ErrInvalidFeature
+	}
+
+	return newDBCache(logger, db, feature, opts...), nil
+}
```

Comment:
Returning `EncryptionKeycache` or `SigningKeycache` is a nice way to prevent the caller from encrypting & signing with the same key (which is a big no-no).

You could still typecast this into the other type. Do you think adding `isEncryptionKeyFeature` and `isSigningKeyFeature` to the functions that return the signing/encrypt key would be excessive? Just to futher protect from any misuse.

Reply:
I was doing the checks originally but I removed it because it feels like it's so outside the spirit of the API it's not even worth accounting for. I can add it back though if you think it's better.

Reply:
Your call, I felt the check was cheap

Reply:
Done